src/compile/data/bin.ts

```diff
@@ -16,57 +16,62 @@ function numberFormatExpr(expr: string, format: string) {
+    const format = (model.axis(channel) || model.legend(channel) || {}).format ||
+      model.config.numberFormat;
 
-      if (!binTrans.maxbins && !binTrans.step) {
-        // if both maxbins and step are not specified, need to automatically determine bin
-        binTrans.maxbins = autoMaxBins(channel);
-      }
+    const startField = field(fieldDef, {datum: true, binSuffix: 'start'});
+    const endField = field(fieldDef, {datum: true, binSuffix: 'end'});
 
-      transform.push(binTrans);
-
-      const hasDiscreteDomainOrHasLegend = hasDiscreteDomain(model.scale(channel).type) || model.legend(channel);
-      if (hasDiscreteDomainOrHasLegend) {
-        // read format from axis or legend, if there is no format then use config.numberFormat
-        const format = (model.axis(channel) || model.legend(channel) || {}).format ||
-          model.config.numberFormat;
-
-        const startField = field(fieldDef, {datum: true, binSuffix: 'start'});
-        const endField = field(fieldDef, {datum: true, binSuffix: 'end'});
+    transform.push({
+      type: 'formula',
+      as: field(fieldDef, {binSuffix: 'range'}),
+      expr: `${numberFormatExpr(startField, format)} + ' - ' + ${numberFormatExpr(endField, format)}`
+    });
+  }
+}
 
-        transform.push({
-          type: 'formula',
-          as: field(fieldDef, {binSuffix: 'range'}),
-          expr: `${numberFormatExpr(startField, format)} + ' - ' + ${numberFormatExpr(endField, format)}`
-        });
+function parse(model: Model): Dict<VgTransform[]> {
+  return model.reduceFieldDef(function(binComponent: Dict<VgTransform[]>, fieldDef: FieldDef, channel: Channel) {
+    const fieldDefBin = model.fieldDef(channel).bin;
+    if (fieldDefBin) {
+      // Make this bin variable always an object
+      // and extend it with maxbins if needed here.
+      const bin: Bin = isBoolean(fieldDefBin) ? {} : {...fieldDefBin};
+      if (!bin.maxbins && !bin.steps) {
+        bin.maxbins = autoMaxBins(channel);
+      }
+      const key = hash(bin) + '_' + fieldDef.field;
+      if (binComponent[key]) {
+        checkDiscreteDomainOrLegend(model, binComponent[key], fieldDef, channel);
+      } else {
+        const transform: VgTransform[] = [];
+        const extentSignal = model.getName(key + '_extent');
+        const binTrans: VgTransform = {
+          ...{
+            type: 'bin',
+            field: fieldDef.field,
+            as: [field(fieldDef, {binSuffix: 'start'}), field(fieldDef, {binSuffix: 'end'})]
+          },
+          ...bin,
+          // add extent if it's not specified
+          ...(!bin.extent ? {extent: {signal: extentSignal}} : {})
+        };
+        if (!bin.extent) {
+          transform.push({
+            type: 'extent',
+            field: fieldDef.field,
+            signal: extentSignal
+          });
+        }
+        transform.push(binTrans);
+
+        checkDiscreteDomainOrLegend(model, transform, fieldDef, channel);
```

Comment:
You can refactor code such that you only need to call this method once at the end.